c10/util/Metaprogramming.h

```diff
@@ -5,49 +5,75 @@
+ * 
  * Access information about result type or arguments from a function type.
  * Example:
  * using A = function_traits<int (float, double)>::return_type // A == int
  * using A = function_traits<int (float, double)>::parameter_types::tuple_type // A == tuple<float, double>
  */
-template<class Func> struct function_traits {
-  static_assert(!std::is_same<Func, Func>::value, "In function_traits<Func>, Func must be a plain function type.");
-};
-template<class Result, class... Args>
-struct function_traits<Result (Args...)> {
-  using func_type = Result (Args...);
-  using return_type = Result;
-  using parameter_types = typelist::typelist<Args...>;
-  static constexpr auto number_of_parameters = sizeof...(Args);
-};
 
-/**
- * infer_function_traits: creates a `function_traits` type for a simple
- * function (pointer) or functor (lambda/struct). Currently does not support
- * class methods.
- */
+// Modified from https://stackoverflow.com/questions/7943525/is-it-possible-to-figure-out-the-parameter-type-and-return-type-of-a-lambda
 
-template <typename Functor>
-struct infer_function_traits {
-  using type = function_traits<c10::guts::detail::strip_class_t<decltype(&Functor::operator())>>;
+// Fallback, anything with an operator()
+template <typename T>
+struct function_traits : public function_traits<decltype(&T::operator())> {
 };
 
-template <typename Result, typename... Args>
-struct infer_function_traits<Result (*)(Args...)> {
-  using type = function_traits<Result(Args...)>;
+// Pointers to class members that are themselves functors.
+// For example, in the following code:
+// template <typename func_t>
+// struct S {
+//     func_t f;
+// };
+// template <typename func_t>
+// S<func_t> make_s(func_t f) {
+//     return S<func_t> { .f = f };
+// }
+//
+// auto s = make_s([] (int, float) -> double { /* ... */ });
+//
+// function_traits<decltype(&s::f)> traits;
+template <typename ClassType, typename T>
+struct function_traits<T ClassType::*> : public function_traits<T> {
 };
 
-template <typename Result, typename... Args>
-struct infer_function_traits<Result (Args...)> {
-  using type = function_traits<Result(Args...)>;
+// Const class member functions
+template <typename ClassType, typename ReturnType, typename... Args>
+struct function_traits<ReturnType(ClassType::*)(Args...) const> : public function_traits<ReturnType(Args...)> {
 };
 
+// Reference types
+template <typename T>
+struct function_traits<T&> : public function_traits<T> {};
 template <typename T>
-using infer_function_traits_t = typename infer_function_traits<T>::type;
+struct function_traits<T*> : public function_traits<T> {};
+
+// Free functions
+template <typename ReturnType, typename... Args>
+struct function_traits<ReturnType(Args...)> {
+  using func_type = ReturnType (Args...);
+  using return_type = ReturnType;
+  using parameter_types = typelist::typelist<Args...>;
+
+  // arity is the number of arguments.
+  constexpr static int arity = sizeof...(Args);
```

Comment:
This can cause issues because technically, constexpr static int variables in C++14 that are declared in a class body need to have a definition outside of the class body and while some compilers are lenient, other compilers enforce that. Let's do a constexpr function instead:

```
constexpr static int arity() {return sizeof...(Args);}
```

Reply:
fixed